Transition plugin: only add prefixes to value if not added yet

ajoslin opened the pull request:
This is obviously a dumb implementation, and it probably isn't good enough. I just wanted to start a discussion.

How can I improve this to make sure it works in every case? Should I be checking that `/WebkitTransition(Property)?/` already exists in the keys?

rofrischmann commented:
Hmm. I think the best solution would be to add another utility function called `isPrefixedValue` or sth. similar and to use it every time the value gets checked/prefixed. We should only process a value if it has not already been prefixed.

`isPrefixedValue` would basically just look sth. like this
```javascript
const isPrefixedValue = value => value.match(/-webkit-|-moz-|-ms-/) !== null 
```

ajoslin commented:
:+1: cool, just force pushed with that implementation. How's it look now?

rofrischmann commented:
Yeah seems to work exactly as it should. I think I am ok with it but I will check it in detail tomorrow after my exam (3am right now here in Germany :P).
Have you tested this issue with something else, e.g. `calc()`? I guess we will also face the issue trying to prefix e.g. `width: -webkit-calc(100% - 10px)`.

ajoslin commented:
No huge rush on this, it's not super urgent for me -- I'm just using `transition: all` until it works.

I'll add some unit tests for a few more prefixed properties soon.

Good luck on the exams! 💯

ajoslin commented:
Ok, turns out there were some more major changes needed. I split out into many commits to make it easier to read.

- API change: prefixed values no longer return arrays. Hiding the values in arrays made it a lot harder to tell if they were prefixed yet, and I didn't see the value in returning an array here. I can change this back if you wish, but then we need to add some `isArray` checking logic.

I fixed calc, gradient and transition all not being idempotent, and made all plugin tests do a double-prefix to.

rtsao commented:
If an array isn't returned, how would one actually use the multiple values? Attempting to use a value containing commas will not work (either turning into CSS or assigning as an inline style), so the only option I see would be to attempt String.split(',') for *every* value, which would probably be a non-trivial performance drain.

rtsao commented:
@rofrischmann It never returned commas. I was just looking at the PR 😉 
For example: https://github.com/rofrischmann/inline-style-prefix-all/pull/9/files#diff-30a2994ff07c8f51474e96d66bcb31a0R57

ajoslin commented:
@rtsao A string value with commas is the only thing that will work as CSS/inline style. An array has never been a valid CSS/style value. I'm not sure what you're saying.

`width: ['calc(100%)', '-webkit-calc(100%)']` is an invalid style.

`width: 'calc(100%),-webkit-calc(100%)` is vlaid.

@rofrischmann I can take it back to being an array, and we can do "array is already prefixed" as a rule, but it needs more thought than that. It's possible the user of this library could, for example, create their own array of `transition-property` values and pass it into prefixAll. We'd still need to prefix those transition values.

ajoslin commented:
I'll make it so if an array passed in already has one or more prefixed values in it, we count it as "already prefixed".

rtsao commented:
@ajoslin Are you sure comma separated values are valid? Comma separation seems to be non-functional, for example:
http://codepen.io/rtsao/pen/ZWmNoo

Arrays obviously aren't valid values, but it makes it easier for consumers to turn them into valid inline styles.

ajoslin commented:
Yes you're right, I edited my above comment after writing because I realized the different functionality of arrays.

ajoslin commented:
OK, it was a lot simpler than anticipated. Squashed it all into one commit again without requiring any API changes.

rofrischmann commented:
@rtsao Oh okay sorry I got that wrong. 
@ajoslin Actually a string value with commas is not valid CSS. You would need to use `;` first of all and each value must follow another `property + ':'`. 
```CSS
/* invalid */
width: calc(100%),-webkit-calc(100%)

/* valid */
width: calc(100%);width:-webkit-calc(100%)
```

The reason why I chose to support the array-returning API is that inline styles do not support alternative values via string. React supported those up to version 14.x as a side-effect, but since they moved to `.createElement` all string values containing fallbacks are no longer supported. 

If you want to generate CSS you just do the `array.join(';' + property + ':')`, but if you are going pure inline e.g. as Radium does, you would need to validate each value to find supported ones.

ajoslin commented:
:+1: sounds great. I added a dependency to `isarray`, which is a 3-line module. I see you have no external deps right now, so if you don't want `isarray` either let me know.

rofrischmann commented on package.json:
```diff
@@ -57,5 +57,8 @@
     "uglifyjs": "^2.4.10"
+  },
+  "dependencies": {
+    "isarray": "^1.0.0"
```
Do we actually need this? I was used to `Array.isArray(arr)`, but I am not sure if that's supported by all major browsers.

rtsao replied:
It's supported for the browsers that this module supports: http://kangax.github.io/compat-table/es5/#test-Array_methods_Array.isArray

rofrischmann replied:
@ajoslin FYI :P I guess we can just use `Array.isArray` for this.

ajoslin replied:
:+1: removed.
